### PR TITLE
Profile photo upload

### PR DESCRIPTION
**Pull request by ghost**
Profile photo can be uploaded by user profile form. See commit messages for more info.

### CODE REVIEW

**Comment by ghost** [comment]
As discussed in down the track this could be an option - https://django-storages.readthedocs.org/en/latest/

Right now I am changing this FileField to an Image field and trying to use the file system for storage.

**Comment by benjaoming** [comment]
Fixes #58

**Comment by benjaoming** [comment]
@jambroo this looks good, IMO we can solve the challenge of storage later by adding a different default storage backend, see [DEFAULT_FILE_STORAGE](https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-DEFAULT_FILE_STORAGE)

Could you rebase against current master?

File: eggplant/profiles/views.py [hunk 1]
```diff
@@ -103,12 +105,17 @@ def get_initial(self):
 
     def form_valid(self, form):
         user_id = self.request.user.id
+        photo = self.request.FILES.get('photo')
```
Comment (ghost) [inline eggplant/profiles/views.py]: I get an InMemoryUploadedFile here but can't seem to save it on the UserProfile below. Any ideas?

File: eggplant/profiles/views.py [hunk 2]
```diff
@@ -91,6 +92,7 @@ class Profile(LoginRequiredMixin, FormView):
 
     def get_object(self, queryset=None):
         self.objects = UserProfile.objects.get(user_id=self.request.user.id)
```
Comment (benjaoming) [inline eggplant/profiles/views.py]: should this be self.object?

File: eggplant/profiles/views.py [hunk 3]
```diff
@@ -103,12 +105,17 @@ def get_initial(self):
...
     def form_valid(self, form):
         user_id = self.request.user.id
+        photo = self.request.FILES.get('photo')
+
         User.objects.filter(id=user_id)\
             .update(first_name=form.cleaned_data['first_name'],
                     last_name=form.cleaned_data['last_name'])
         del form.cleaned_data['first_name']
         del form.cleaned_data['last_name']
+        del form.cleaned_data['photo']
```
Comment (benjaoming) [inline eggplant/profiles/views.py]: Maybe this line is problematic? Actually, why are these deleted from the cleaned_data object? I think it should be part of forms.cleaned_data which is set for the UserProfile. You could use `self.object.field = form.cleaned_data[...]` and then `self.object.save()`....

File: eggplant/profiles/models.py [hunk 4]
```diff
@@ -4,7 +4,16 @@
 from django.dispatch.dispatcher import receiver
 from django.db.models.signals import post_save
 from django.contrib.auth.models import User
+from django.core.files.storage import FileSystemStorage
+import os
 
+# Prepare profile photo dir
+profile_photo_dir = settings.BASE_DIR + "/media/profile/"
```
Comment (pawciobiel) [inline eggplant/profiles/models.py]: In core.utils there is a function called `generate_upload_path` it may make sense to use it to have random file names.

File: eggplant/profiles/models.py [hunk 5]
```diff
@@ -4,7 +4,16 @@
...
+
+if not os.path.exists(profile_photo_dir):
+    os.makedirs(profile_photo_dir)
+
+profile_photo_fs = FileSystemStorage(location=profile_photo_dir)
```
Comment (pawciobiel) [inline eggplant/profiles/models.py]: During the hack-day I've mention this just let you know about `storage` kwarg. but we could just use settings.DEFAULT_FILE_STORAGE for now.

File: eggplant/profiles/views.py (hunk 3 above)
Comment (pawciobiel) [inline eggplant/profiles/views.py]: @jambroo You can't use `.update` because it runs on profile table only. @benjaoming mentioned above use `save`.

File: eggplant/profiles/views.py (hunk 2 above)
Comment (pawciobiel) [inline eggplant/profiles/views.py]: I had a look at `django.views.generic.edit` and I don't think this matters so it looks like a typo to me.

Simple typo and we keep wondering... -  there you go, CBV for you.

File: eggplant/profiles/views.py [hunk 6]
```diff
@@ -1,13 +1,15 @@
 import logging
-
+import base64
```
Comment (pawciobiel) [inline eggplant/profiles/views.py]: lint will throw "unused import"

File: eggplant/profiles/views.py (hunk 2 above)
Comment (benjaoming) [inline eggplant/profiles/views.py]: Regardless of CBV patterns and the parent class' contract, this would be a misleading typo.

File: eggplant/profiles/models.py (hunk 4 above)
Comment (benjaoming) [inline eggplant/profiles/models.py]: :+1: great util!

File: eggplant/profiles/views.py (hunk 6 above)
Comment (ghost) [inline eggplant/profiles/views.py]: Okay I will get rid of this

File: eggplant/profiles/models.py (hunk 5 above)
Comment (ghost) [inline eggplant/profiles/models.py]: I have taken this code out and it should use whatever the default is. Not sure if the folder needs to be made on the live server or anything like that :)

File: eggplant/profiles/views.py (hunk 3 above)
Comment (ghost) [inline eggplant/profiles/views.py]: Good point - using self.objects.photo = form.cleaned_data['photo'] made it work

**Comment by ghost** [comment]
@pawciobiel @benjaoming - can this be merged?

**Comment by benjaoming** [comment]
@jambroo I think you should change `self.objects` to `self.object` because it's sort of odd with the pluralization :) Otherwise :+1:

**Comment by ghost** [comment]
@benjaoming Okay that is done :) Couldn't get TravisCI to re-test it though..

**Comment by benjaoming** [comment]
Oh one more thing...

I have a feeling that updating first name and last name won't work because firstly, we do an `update()` which spawns an SQL UPDATE query... but then after that, we use an old instance `self.object` which still contains the old values of `first_name` and `last_name` -- could you investigate if changing these are still possible? Perhaps it's better to just remove the `update` pattern and add first_name and last_name in the below pattern...

**Comment by ghost** [comment]
I think this is creeping out a bit so it is no longer associated with the profile photo but I can say for certain the first_name, last_name and middle_names are being saved as the test_profile unit test confirms. It is a little weird the first and last name are on the User model and the middle name is on the Profile model but maybe it can be refactored in a later PR.

**Comment by benjaoming** [comment]
Okay, it's cool... let's get this merged :)
